docs/docs/09-rfcs/2023-03-08-wing-preview-environment.md

```diff
@@ -0,0 +1,102 @@
+---
+title: Wing Cloud Preview Environments
+description: This document describes the Wing Cloud Preview Environments feature.
+---
+
+# #3696 - Wing Cloud Preview Environments
+
+- **Author(s)**: @ainvoner, @skyrpex
+- **Submission Date**: 2023-08-03
+- **Stage**: Proposal
+- **Stage Date**: 2023-08-03
+
+This document describes the Wing Cloud Preview Environments feature.
+
+## Introduction
+The **Wing Cloud Preview Environments** project is designed to achieve two key objectives. Firstly, it aims to offer a preview environment solution for those interested in using winglang.
+Secondly, it seeks to enhance the winglang ecosystem and its toolchain by building a production application using the language.
+
+## Requirements
+Create the first production feature of Wing Cloud - Wing cloud Preview Environments.
+Build and Deploy wing applications to a Wing Console simulated environment.
+Allow developers to iterate with their teammates using a shared Wing Console instance.
+Seamlessly integrating with GitHub.
+
+### Developer Experience
+A developer using winglang has the capability to integrate preview environments with his/her repository.
+
+#### Sign Up and Installation
+In order to use Wing Cloud Preview Environments, you should sign up to Wing Cloud.
+The signup and installation process is straightforward using GitHub application:
+
+1. Goto https://wing.cloud website and click "Deploy with Wing Cloud"
+2. Complete GitHub authentication.
+3. Grant repository access permission.
+4. You are now signed up and ready to use Wing Cloud Preview Environments.
+
+#### Preview Environment for production branch
+For the production branch, there is a single preview environment that will constantly be up-to-date with the latest code committed.
```

Comment:
how will URLs look like?

Ideally something like:

```
https://wing.cloud/gh-account/gh-repo/gh-branch
```

e.g.


https://wing.cloud/winglang/docsite/main

---

this should be the same for PRs (because they must be branched), no?

Reply:
I thought of using subdomains for wing.cloud:
1. Production branch environment url structure is: `https://<gh-repository>-<production-branch-name>.wing.cloud.app`
2. PR preview environments urls structure is: `https://<gh-repository>-<gh-branch>-<pr-number>-<entry-point>.wing.cloud.app`

Reply:
Why not normal API routes? What's the benefit of using subdomains? I think normal URL routes are more suitable for this need.

Reply:
Using subdomains is ~preferred~ mandatory here, security wise.

The preview environments will run code created by users. If we serve them under wing.cloud, we expose us to [XSRF attacks](https://en.wikipedia.org/wiki/Cross-site_request_forgery).

Even if this preview environment is served under the Console App control, it's very likely that users may be able to abuse it. Attacks that would be possible are:

- Gathering information from the user previewing the environment
- Executing calls to wing.cloud in the name of this user

I think we can run the subdomains under `*.wing.cloud` but I need to do some research first... there's some HTTP headers that can protect the main site from the subdomains. But the safest option is to use a totally different domain, something like `*.wing-preview.cloud`, so there's no open doors for XSRF.

Reply:
@skyrpex isn't wing.cloud.app is a different domain than wing.cloud? isn't it enough?

Reply:
I am okay to use subdomains if that's the only way to defend against attacks but it should be under the wing.cloud parent

Reply:
but we need a different domain... accessing our wing.cloud auth cookie for example would be possible from preview environments if we will use sub domains of wing.cloud.

Reply:
> I am okay to use subdomains if that's the only way to defend against attacks but it should be under the wing.cloud parent

I'm afraid we'll have to use a different domain altogether. There's multiple security risks involved...

Just to give you an example:
- After signing in to `https://wing.cloud`, there'll be an secure HTTP-only cookie on `wing.cloud` used for authentication
- If the preview environments are `*.wing.cloud`, the browser will send this cookie to them and they'll be able to read your session credentials and stuff on your behalf

There are security measures we could take to prevent attacks, but it'd be like if the whole Wing Cloud's security was hanging from a thread. The smallest mistake would become a security breach. We need to go for something like `*.wingcloud.app`.